File: src/car/Control.py
import RPi.GPIO as GPIO
import pigpio
# This class controls the two motors on the car

class Control:
    def __init__(self):
        # set up PWM pigpio pins
        self.pi = pigpio.pi()
        self.pin1A = 12
        self.pin1B = 16
        self.pin2A = 22
        self.pin2B = 25
        # set up gpio output pins for motors one and two
        GPIO.setmode(GPIO.BCM)
        # front motors
        self.IN1_MOTOR1_PIN = 17
        self.IN2_MOTOR1_PIN = 27   
        self.IN1_MOTOR2_PIN = 23
        self.IN2_MOTOR2_PIN = 24
        # back motors
        self.IN1_MOTOR3_PIN = 26
        self.IN2_MOTOR3_PIN = 13
        self.IN1_MOTOR4_PIN = 6
        self.IN2_MOTOR4_PIN = 5
        # set up
        GPIO.setup(self.IN1_MOTOR1_PIN, GPIO.OUT, initial=GPIO.LOW)
        GPIO.setup(self.IN2_MOTOR1_PIN, GPIO.OUT, initial=GPIO.LOW)             
        GPIO.setup(self.IN1_MOTOR2_PIN, GPIO.OUT, initial=GPIO.LOW)
        GPIO.setup(self.IN2_MOTOR2_PIN, GPIO.OUT, initial=GPIO.LOW) 
        GPIO.setup(self.IN1_MOTOR3_PIN, GPIO.OUT, initial=GPIO.LOW)
        GPIO.setup(self.IN2_MOTOR3_PIN, GPIO.OUT, initial=GPIO.LOW)
        GPIO.setup(self.IN1_MOTOR4_PIN, GPIO.OUT, initial=GPIO.LOW)
        GPIO.setup(self.IN2_MOTOR4_PIN, GPIO.OUT, initial=GPIO.LOW)
        self.pi.set_PWM_frequency(16, 100)       # 100Hz frequency
        self.pi.set_PWM_dutycycle(16, 128)       # 50% duty cycle (128 out of 255)
        self.pi.set_PWM_frequency(12, 100)       # 100Hz frequency
        self.pi.set_PWM_dutycycle(12, 128)       # 50% duty cycle (128 out of 255)

        self.pi.set_PWM_frequency(22, 100)       # 100Hz frequency
        self.pi.set_PWM_dutycycle(22, 128)       # 50% duty cycle (128 out of 255)
        self.pi.set_PWM_frequency(25, 100)       # 100Hz frequency
        self.pi.set_PWM_dutycycle(25, 128)       # 50% duty cycle (128 out of 255)

    def motorsForward(self):
        # move motor one forward
        GPIO.output(self.IN1_MOTOR1_PIN, GPIO.HIGH)
        GPIO.output(self.IN2_MOTOR1_PIN, GPIO.LOW)
        # move motor two forward
        GPIO.output(self.IN1_MOTOR2_PIN, GPIO.HIGH)
        GPIO.output(self.IN2_MOTOR2_PIN, GPIO.LOW)
        # move motor three forward
        GPIO.output(self.IN1_MOTOR3_PIN, GPIO.LOW)
        GPIO.output(self.IN2_MOTOR3_PIN, GPIO.HIGH)
        # move motor four forward
        GPIO.output(self.IN1_MOTOR4_PIN, GPIO.HIGH)
        GPIO.output(self.IN2_MOTOR4_PIN, GPIO.LOW)

    def turnLeft(self):
        # turn motor 1 off
        GPIO.output(self.IN1_MOTOR1_PIN, GPIO.LOW)
        GPIO.output(self.IN2_MOTOR1_PIN, GPIO.LOW)

    def turnRight(self):
        # turn motor 2 off
        GPIO.output(self.IN1_MOTOR2_PIN, GPIO.LOW)
        GPIO.output(self.IN2_MOTOR2_PIN, GPIO.LOw)

    def motorsStop(self):
        # stop motor one
        GPIO.output(self.IN1_MOTOR1_PIN, GPIO.LOW)
        GPIO.output(self.IN2_MOTOR1_PIN, GPIO.LOW)
        # stop motor two
        GPIO.output(self.IN1_MOTOR2_PIN, GPIO.LOW)
        GPIO.output(self.IN2_MOTOR2_PIN, GPIO.LOW)
        # stop motor three
        GPIO.output(self.IN1_MOTOR3_PIN, GPIO.LOW)
        GPIO.output(self.IN2_MOTOR3_PIN, GPIO.LOW)
        # stop motor four
        GPIO.output(self.IN1_MOTOR4_PIN, GPIO.LOW)
        GPIO.output(self.IN2_MOTOR4_PIN, GPIO.LOW)

    def cleanUp(self):
        # Clean up GPIO settings
        GPIO.cleanup()
        self.pi.set_PWM_dutycycle(12, 0)
        self.pi.set_PWM_dutycycle(16, 0)
        self.pi.set_PWM_dutycycle(22, 0)
        self.pi.set_PWM_dutycycle(25, 0)
        self.pi.stop()
        # self.pwm.stop() 

File: src/car/Control.c
#include <stdio.h>
#include "Control.h"
#include <pigpio.h>

void GPIOSetup() {
    // set up outputs
    gpioSetMode(PIN1_MOTOR1, PI_OUTPUT);
    gpioSetMode(PIN2_MOTOR1, PI_OUTPUT);
    gpioSetMode(PIN1_MOTOR2, PI_OUTPUT);
    gpioSetMode(PIN2_MOTOR2, PI_OUTPUT);
    gpioSetMode(PIN1_MOTOR3, PI_OUTPUT);
    gpioSetMode(PIN2_MOTOR3, PI_OUTPUT);
    gpioSetMode(PIN1_MOTOR4, PI_OUTPUT);
    gpioSetMode(PIN2_MOTOR4, PI_OUTPUT);
    // set status
    gpioWrite(PIN1_MOTOR1, PI_LOW);
    gpioWrite(PIN1_MOTOR1, PI_LOW);
    gpioWrite(PIN1_MOTOR2, PI_LOW);
    gpioWrite(PIN1_MOTOR2, PI_LOW);
    gpioWrite(PIN1_MOTOR3, PI_LOW);
    gpioWrite(PIN1_MOTOR3, PI_LOW);
    gpioWrite(PIN1_MOTOR4, PI_LOW);
    gpioWrite(PIN1_MOTOR4, PI_LOW);
}

void pwmDutyCycle(int cycleNum) {
    gpioPWM(PIN_1A, cycleNum);
    gpioPWM(PIN_1B, cycleNum);
    gpioPWM(PIN_2A, cycleNum);
    gpioPWM(PIN_2B, cycleNum);
}

void motorsStop() {
    pwmDutyCycle(0);
    gpioWrite(PIN1_MOTOR1, PI_LOW);
    gpioWrite(PIN2_MOTOR1, PI_LOW);
    gpioWrite(PIN1_MOTOR2, PI_LOW);
    gpioWrite(PIN2_MOTOR2, PI_LOW);
    gpioWrite(PIN1_MOTOR3, PI_LOW);
    gpioWrite(PIN2_MOTOR3, PI_LOW);
    gpioWrite(PIN1_MOTOR4, PI_LOW);
    gpioWrite(PIN2_MOTOR4, PI_LOW);
}

void cleanup() {
    // stop motors
    motorsStop();    
    // clean gpio
    gpioTerminate();
}

void pwmFrequency(int freqNum) {
    gpioSetPWMfrequency(PIN_1A, freqNum);
    gpioSetPWMfrequency(PIN_1B, freqNum);
    gpioSetPWMfrequency(PIN_2A, freqNum);
    gpioSetPWMfrequency(PIN_2B, freqNum);
}

void PWMSetup() {
    gpioSetMode(PIN_1A, PI_OUTPUT);
    gpioSetMode(PIN_1B, PI_OUTPUT);
    gpioSetMode(PIN_2A, PI_OUTPUT);
    gpioSetMode(PIN_2B, PI_OUTPUT);
}

int convertThrottle(int throttle) {
    int convertedThrottle;
    if (throttle == 0) {
    	convertedThrottle = 0;
    } else if (throttle == 1) {
    	convertedThrottle = 128;
    } else if (throttle == 2) {
	convertedThrottle = 192;
    } else if (throttle == 3) {
	convertedThrottle = 224;	    
    } else {
	convertedThrottle = 255;
    }
    return convertedThrottle;
}

void motorsForward(int throttle) {
    int convertedThrottle = convertThrottle(throttle);
    pwmDutyCycle(convertedThrottle);
    gpioWrite(PIN1_MOTOR1, PI_HIGH);
    gpioWrite(PIN2_MOTOR1, PI_LOW);
    gpioWrite(PIN1_MOTOR2, PI_HIGH);
    gpioWrite(PIN2_MOTOR2, PI_LOW);
    gpioWrite(PIN1_MOTOR3, PI_HIGH);
    gpioWrite(PIN2_MOTOR3, PI_LOW);
    gpioWrite(PIN1_MOTOR4, PI_HIGH);
    gpioWrite(PIN2_MOTOR4, PI_LOW);
}

void turnRight() {
    // turn motor 2 off
    gpioWrite(PIN1_MOTOR2, PI_LOW);
    gpioWrite(PIN2_MOTOR2, PI_LOW);
}

void turnLeft() {
    // turn motor 1
    gpioWrite(PIN1_MOTOR1, PI_LOW);
    gpioWrite(PIN2_MOTOR1, PI_LOW);
}

int setup() {
    printf("Set up Car process start\n");
    // init pigpio
    if (gpioInitialise() < 0) {
        printf("pigpio initialization failed.\n");
        return 1;
    }

    GPIOSetup();
    PWMSetup();
    pwmFrequency(100);
    pwmDutyCycle(0);
    return 0;
}
